---
title:  "COG-UK Mutations"
output: html_notebook
params: 
  mutations_csv:  "cog_global_2021-02-25_mutations.csv"
  global_csv:     "cog_global_2021-02-25_public.csv"
  consortium_csv: "cog_global_2021-02-25_consortium.csv"
  spike_csv:      "spike_escape_info.csv"
---
```{r libs}
library(tidyverse)
library(magrittr)
library(wesanderson)
library(RColorBrewer)
library(lubridate)
```

## Consortium
```{r consortium_read}
consortium_uk <- 
  read_csv((params$consortium_csv), col_types = cols(.default = col_character())) %>% 
  select(-cog_id, -gisaid_id, -submission_org_code, -root_sample_id, -(adm2:travel_history), -lineage_support, -del_lineage, -del_introduction) %>% # remove unused columns
  type_convert %>% 
  filter(country == "UK")  %>% 
  distinct # remove duplicates

epi_levels <- min(consortium_uk$epi_week):max(consortium_uk$epi_week) %>% as.character

consortium_uk %<>% 
  mutate(across(c(epi_week, adm1), as_factor)) %>%
  mutate(epi_week = fct_expand(epi_week, epi_levels) %>% fct_inseq) %T>% 
  write_rds("consortium_uk.rds")

sample_date_28 <- max(consortium_uk$sample_date) - days(27) # calculate 28 day period up to and including latest sample date

consortium_uk
```

## Mutations
```{r read}
mutations <-   
  read_csv(params$mutations_csv, col_types = cols(.default = col_character())) %>% 
  type_convert %>% 
  distinct %>% # get rid of identical variant calls
  separate_rows(variants, sep = '\\|') %>% 
  separate(variants, into = c("gene", "variant"), sep = ':') %>% 
  drop_na() %>%
  filter(gene != "synSNP") %>%
  mutate(position = parse_number(variant))

public <- 
  read_csv(params$global_csv, col_types = cols(.default = col_character())) %>% 
  type_convert %>% 
  select(-cog_id, -gisaid_id, -(pillar_2:travel_history), -lineage_support, -del_lineage, -del_introduction) %>% 
  distinct # remove duplicates

mutations %<>% inner_join(public)

mutations_uk <- 
  mutations %>%
  filter(country == "UK") %>% 
  mutate(across(c(epi_week, gene, position, variant, adm1), as_factor)) %>%
  mutate(epi_week = fct_expand(epi_week, epi_levels) %>% fct_inseq) %T>% 
  write_rds(file = "mutations_uk.rds")

mutations_uk
```

## Mutation / REF counts
```{r}
#### UK ### 
# TODO calculate UK counts by summing nation counts
genes_positions <- 
  mutations_uk %>% 
  distinct(gene, position) # gene / position combinations actually found in mutations

sequences_by_week <- 
  consortium_uk %>% 
  count(epi_week, name = "n_sequences")

positions_by_week <- 
  mutations_uk %>%
  count(epi_week, gene, position, .drop = FALSE, name = "n_variant_sequences") # count all possible epi_week/gene/position combinations 

reference_counts <- 
  positions_by_week %>%
  inner_join(genes_positions) %>% # remove gene / position combinations not found in mutations
  inner_join(sequences_by_week) %>% 
  mutate(n = n_sequences - n_variant_sequences, variant = "WT", .keep = "unused") %>% # calculate reference / other
  filter(n > 0) %>%  # remove zero WT entries (where all sequences are variants for an epi_week / position combo)
  mutate(adm1 = "UK")

mutation_counts <- 
  mutations_uk %>%
  count(epi_week, gene, position, variant) %>% 
  mutate(adm1 = "UK")

#### Nations ####
levels_adm1 <- c(Scotland = "UK-SCT",
                 Wales = "UK-WLS", 
                 England = "UK-ENG", 
                 Northern_Ireland = "UK-NIR") # adm1 factor levels from consortium

sequences_by_week_nations <-
  consortium_uk %>%
  count(epi_week, adm1, name = "n_sequences")

positions_by_week_nations <-
  mutations_uk %>%
  mutate(adm1 = fct_recode(adm1, !!!levels_adm1)) %>%  # recode nation names from consortium
  count(epi_week, gene, position, adm1, .drop = FALSE, name = "n_variant_sequences") # count all possible epi_week/gene/position/nation combinations

reference_counts_nations <- 
  positions_by_week_nations %>%
  inner_join(genes_positions) %>% # remove gene / position combinations not found in mutations
  inner_join(sequences_by_week_nations) %>% 
  mutate(n = n_sequences - n_variant_sequences, variant = "WT", .keep = "unused") %>% # calculate reference / other
  filter(n > 0) # remove zero WT entries (where all sequences are variants for an epi_week / position combo)

mutation_counts_nations <- 
  mutations_uk %>%
  mutate(adm1 = fct_recode(adm1, !!!levels_adm1)) %>%  # recode nation names from consortium
  count(epi_week, gene, position, variant, adm1)

# TODO calculate UK counts by summing the nation counts
mutation_reference_counts <- 
  bind_rows(mutation_counts, reference_counts, mutation_counts_nations, reference_counts_nations) %>% 
  mutate(across(variant, as_factor)) %T>%
  write_rds("mutation_reference_counts.rds")

mutation_reference_counts
```
### Total sequences per epidemic week

```{r consortium_counts}
n_uk <- 
  consortium_uk %>% 
  # filter(sample_date >= 2020-11-13) %>% 
  # filter(epi_week < 50) %>%
  group_by(epi_week) %>%
  summarise(sequences = n(), D614G = sum(d614g == "G"), A222V = sum(a222v == "V"), 
            N439K = sum(n439k == "K"), N501Y = sum(n501y == "Y"), DEL_69_70 = sum(del_21765_6 == "del")) #%>%
            # mutate_at(vars(D614G:DEL_69_70), cumsum) # cumulative counts
            # mutate_at(vars(sequences:DEL_69_70), cumsum) # cumulative counts to normalise against cumulative sequences

n_uk
```



```{r}
n_uk %>% 
  select(-epi_week) %>% 
  summarise_all(funs(sum))
```

```{r}
n_uk_28 <- 
  consortium_uk %>% 
  filter(sample_date >= sample_date_28) %>%
  group_by(epi_week) %>%
  summarise(sequences = n(), D614G = sum(d614g == "G"), A222V = sum(a222v == "V"), 
            N439K = sum(n439k == "K"), N501Y = sum(n501y == "Y"), DEL_69_70 = sum(del_21765_6 == "del"))
n_uk_28
```
```{r}
n_uk_28 %>% 
  select(-epi_week) %>% 
  summarise_all(funs(sum))
```


```{r}
theme_set(theme_classic() + theme(legend.position = "none"))

pal <- wes_palette("Cavalcanti1", type = "discrete")
update_geom_defaults("point", list(color = pal[1]))
update_geom_defaults("line", list(color = pal[1]))

 ggplot(n_uk, aes(x = epi_week, y = sequences)) + 
  geom_line() + geom_point() + labs(x = "Epidemic week", y = "Sequences", title = "Sequences per Week")

 ggsave("Sequences.png")
```
```{r}
pal <- brewer.pal(n = 5, name = "Set1")
update_geom_defaults("point", list(color = pal[2]))
update_geom_defaults("line", list(color = pal[2]))

 ggplot(n_uk, aes(x = epi_week, y = D614G / sequences)) + 
  geom_line() + geom_point() + labs(x = "Epidemic week", y = "D614G (normalised)", title = "D614G") + ylim(0,1)
 ggsave("D614G.png")

```

```{r}
update_geom_defaults("point", list(color = pal[1]))
update_geom_defaults("line", list(color = pal[1]))

 ggplot(n_uk, aes(x = epi_week, y = A222V / sequences)) + 
  geom_line() + geom_point() + labs(x = "Epidemic week", y = "A222V (normalised)", title = "A222V") + ylim(0,1)
 ggsave("A222V.png")

```
```{r}
update_geom_defaults("point", list(color = pal[4]))
update_geom_defaults("line", list(color = pal[4]))

 ggplot(n_uk, aes(x = epi_week, y = N439K / sequences)) + 
  geom_line() + geom_point() + labs(x = "Epidemic week", y = "N439K (normalised)", title = "N439K") + ylim(0,1)
 ggsave("N439K.png")

```

```{r}
update_geom_defaults("point", list(color = pal[5]))
update_geom_defaults("line", list(color = pal[5]))

 ggplot(n_uk, aes(x = epi_week, y = N501Y / sequences)) + 
  geom_line() + geom_point() + labs(x = "Epidemic week", y = "N501Y (normalised)", title = "N501Y") + ylim(0,1)
 ggsave("N501Y.png")

```
```{r}
update_geom_defaults("point", list(color = pal[3]))
update_geom_defaults("line", list(color = pal[3]))

 ggplot(n_uk, aes(x = epi_week, y = DEL_69_70 / sequences)) + 
  geom_line() + geom_point() + labs(x = "Epidemic week", y = "Deletion 69-70 (normalised)", title = "Deletion 69-70") + ylim(0,1)
ggsave("DEL_69_70.png")
```
```{r}
n_uk_long <- 
  n_uk %>% 
  select(-c(sequences, D614G)) %>% 
  gather(key = "variant", value = "seq_raw", -epi_week) 
n_uk_long
```
```{r}
n_uk_long_norm <- 
  n_uk %>% 
  mutate_at(vars(D614G:DEL_69_70), funs((. / sequences) * 100)) %>%
  select(-c(sequences, D614G)) %>% 
  gather(key = "variant", value = "seq_norm", -epi_week) 
n_uk_long_norm
```

```{r}
theme_set(theme_classic())
n_uk_long %>% 
  ggplot(aes(x = epi_week, y = seq_raw)) + 
  geom_line(aes(color = variant)) + geom_point(aes(color = variant), size = 0.25) +
  labs(x = "Epidemic week", y = "Sequences", title = "Variant Viral Sequences per Week")  +  
  # ylim(0,100) + 
  theme(legend.title = element_blank()) + scale_color_manual(values = pal) 
ggsave("variants_overlay.png")
```
```{r}
theme_set(theme_classic())
n_uk_long_norm %>% 
ggplot(aes(x = epi_week, y = seq_norm)) + 
  geom_line(aes(color = variant)) + geom_point(aes(color = variant), size = 0.25) +
  labs(x = "Epidemic week", y = "% Sequences (normalised)", title = "Variant Viral Sequences per Week (Normalised)")  +  
  # ylim(0,100) + 
  theme(legend.title = element_blank()) + scale_color_manual(values = pal) 
ggsave("variants_overlay_norm.png")
```


```{r}
consortium_uk %>% 
  count(n501y, del_21765_6)
```

```{r}
consortium_uk %>%  
  filter(sample_date >= sample_date_28) %>% 
  count(n501y, del_21765_6)

```
```{r}
consortium_uk %>%  
  count(n439k, del_21765_6)

```
```{r}
consortium_uk %>%  
  filter(sample_date >= sample_date_28) %>% 
  count(n439k, del_21765_6)

```
```{r}

consortium_uk %>% 
  filter(n501y == "Y" & del_21765_6 == "del")  %>% 
  count(lineage) %>% 
  arrange(desc(n))
```
```{r}

consortium_uk %>% 
  filter(sample_date >= sample_date_28 & n501y == "Y" & del_21765_6 == "del")  %>% 
  count(lineage) %>% 
  arrange(desc(n))
```
```{r}
consortium_uk %>% 
  filter(n439k == "K" & del_21765_6 == "del")  %>% 
  count(lineage) %>% 
  arrange(desc(n))
```
```{r}

consortium_uk %>% 
  filter(sample_date >= sample_date_28 & n439k == "K" & del_21765_6 == "del")  %>% 
  count(lineage) %>% 
  arrange(desc(n))
```
```{r}
mutations %>% 
  filter(country == "UK" & gene == "S") %>% 
  summarise(n(), n_distinct(sequence_name))
```
```{r}
mutations %>% filter(country == "UK" & gene == "S") %>% 
  count(variant) %>% 
  filter(n >1) %>% 
  summarise(sum(n))
```


```{r}
mutations %>% 
  filter(country == "UK" & gene == "S") %>% 
  count(variant) %>% filter(n >100) %>% 
  summarise(sum(n))
```
```{r}
 # awk '/S:P681H/
 #  && /S:N501Y/
 #   && /orf1ab:T1001I/
 #   && /orf1ab:A1708D/
 #   && /orf1ab:I2230T/
 #   && /S:A570D/
 #   && /S:T716I/
 #   && /S:S982A/
 #   && /S:D1118H/
 #   && /ORF8:Q27*/
 #   && /ORF8:R52I/
 #   && /ORF8:Y73C/
 #   && /N:D3L/
 #   && /N:S235F/ { print;}' /cephfs/covid/bham/results/phylogenetics/20201215/metadata/cog_global_2020-12-20_mutations.csv
 # 

 # awk '/S:P681H/  && /S:N501Y/ && /orf1ab:T1001I/   && /orf1ab:A1708D/   && /orf1ab:I2230T/   && /S:A570D/   && /S:T716I/   && /S:S982A/   && /S:D1118H/ && /ORF8:Q27*/   && /ORF8:R52I/   && /ORF8:Y73C/   && /N:D3L/   && /N:S235F/ { print;}' cog_global_2020-12-20_mutations.csv 
 # 
 # 
 #  awk '/S:P681H/  && /S:N501Y/ && /orf1ab:T1001I/   && /orf1ab:A1708D/   && /orf1ab:I2230T/   && /S:A570D/   && /S:T716I/   && /S:S982A/   && /S:D1118H/ && /ORF8:Q27*/   && /ORF8:R52I/   && /ORF8:Y73C/   && /N:D3L/   && /N:S235F/ { print;}' cog_global_2020-12-21_mutations.csv > mutations_v_020-12-21.csv


mutations_v <- mutations %>% filter(country == "UK" & lineage == "B.1.1.7"
                     & sample_date >= "2020-11-24"
                     & (
  (gene == "S" & variant == "P681H") | 
  (gene == "S" & variant == "N501Y") | 
  (gene == "orf1ab" & variant == "T1001I") | 
  (gene == "orf1ab" & variant == "A1708D") | 
  (gene == "orf1ab" & variant == "I2230T") | 
  (gene == "S" & variant == "A570D") | 
  (gene == "S" & variant == "T716I") | 
  (gene == "S" & variant == "S982A") | 
  (gene == "S" & variant == "D1118H") | 
  (gene == "ORF8" & variant == "Q27*") |  
  (gene == "ORF8" & variant == "R52I") |
  (gene == "ORF8" & variant == "Y73C") | 
  (gene == "N" & variant == "D3L") | 
  (gene == "N" & variant == "S235F") 
  )) %>% count(sequence_name, lineage) %>% filter(n == 14) #%>% select(-n)

mutations_v
```

```{r}
mutations %>% filter(country == "UK" 
                     # & sample_date >= "2020-11-13" 
                     & (
  (gene == "S" & variant == "K417N") | 
  (gene == "S" & variant == "E484K") | 
  (gene == "S" & variant == "N501Y") 
  )) %>% count(sequence_name) %>% filter(n == 3)


```
```{r cluster5}
# Cluster 5 is Y453F, 69-70del, I692V and M1229I
# 
mutations %>% filter(country == "UK" 
                     # & sample_date >= "2020-11-13" 
                     & (
  (gene == "S" && variant == "Y453F") | 
  (gene == "S" & variant == "I692V") | 
  (gene == "S" & variant == "M1229I") 
  )) %>% count(sequence_name) %>% filter(n == 3)


```


```{r deletions_v}
# 21991-21993, 11288-11296, 21765 (del69-70)
deletions <- 
  read_tsv("uk_deletions.txt") %>% 
  separate_rows(samples, sep = '\\|')

deletions_v <- 
  deletions %>% 
  filter(ref_start %in% c(21991, 11288, 21765)) %>% 
  count(samples) %>% 
  filter(n == 3) %>%
  select(-n)

deletions_v
```
## Database
```{r database}
mutations_s_uk <- 
  mutations_uk %>% 
  filter(gene == "S") %>% 
  mutate(across(c(variant, position), fct_drop))

lineage_distinct <- 
  mutations_s_uk %>% 
  group_by(position, variant) %>% 
  summarise(n_lineages = n_distinct(lineage), .groups = "drop") 

lineages <- 
  mutations_s_uk %>% 
  distinct(position, variant, lineage) %>% 
  group_by(position, variant) %>% 
  arrange(position, variant, lineage, .by_group = TRUE) %>% 
  summarise(lineages = toString(lineage), .groups = "drop")

## UK and nations
# All sample dates
variants_uk <-
  mutations_s_uk %>% 
  group_by(position, variant) %>% 
  summarise(UK = n(), earliest = min(sample_date), .groups = "drop")

variants_nations <- 
  mutations_s_uk %>% 
  count(position, variant, adm1) %>% 
  pivot_wider(names_from = adm1, values_from = n)

# Sample dates in last 28 days
variants_uk_28 <- 
  mutations_s_uk %>% 
  filter(sample_date >= sample_date_28) %>% 
  count(position, variant, name = "UK_28")

variants_nations_28 <-
  mutations_s_uk %>% 
    filter(sample_date >= sample_date_28) %>% 
    count(position, variant, adm1) %>% 
    pivot_wider(names_from = adm1, values_from = n) %>% 
    rename_with(~paste0(., "_28"), .cols = starts_with("UK-"))

database <- 
  read_csv(params$spike_csv, col_types = cols(.default = col_character())) %>%
  type_convert %>% 
  mutate(across(c(mutation, position), as_factor)) %>% 
  mutate(anchor = str_c("<a href='", doi, "'target='_blank'>", citation,"</a>"), .keep = "unused") %>% # hyperlink to citation DOI
  group_by(position, mutation) %>% 
  summarise(escape = escape %>% unique %>% sort %>% toString, 
            anchor = anchor %>% unique %>% str_c(collapse = "<br>"), # line breaks between hyperlinks
            mab = mab %>% any,
            plasma = plasma %>% any,
            vaccine_sera = vaccine_sera %>% any,
            support = support %>% unique %>% toString,
            domain = domain %>% unique %>% toString, 
            .groups = "drop")

join_cols <- c("mutation" = "variant", "position" = "position")

database %<>% 
  full_join(lineage_distinct, by = join_cols) %>% 
  full_join(lineages, by = join_cols) %>% 
  full_join(variants_uk,  by = join_cols) %>% 
  full_join(variants_nations,  by = join_cols) %>% 
  full_join(variants_uk_28,  by = join_cols) %>% 
  full_join(variants_nations_28,  by = join_cols) 

database %<>%
  rename(
    `# Global Lineages associated with` = n_lineages, 
    `Global Lineages` = lineages, 
    
    `numSeqs UK` = UK,
    `numSeqs Eng` = `UK-ENG`,
    `numSeqs NI` = `UK-NIR`,
    `numSeqs Scotland` = `UK-SCT`,
    `numSeqs Wales` = `UK-WLS`,

    `numSeqs UK 28 days` = UK_28,
    `numSeqs Eng 28 days` = `UK-ENG_28`,
    `numSeqs Scotland 28 days` = `UK-SCT_28`,
    `numSeqs Wales 28 days` = `UK-WLS_28`, 
    `numSeqs NI 28 days` = `UK-NIR_28`) %>% 
  mutate(across(starts_with( c("numSeqs", "#") ), ~replace_na(.x, 0L))) %>% 
  mutate(across(c(support, domain), as_factor)) %>% 
  mutate(across(c(position, mutation), as.character)) %>% 
  mutate(across(position, as.integer)) %>% 
  arrange(position, mutation) %>% 
  mutate(across(mutation, as_factor))

database %T>% write_rds("database.rds")
```

```{r}
consortium_v <- 
  consortium_uk %>% 
  filter(n501y == "Y" & 
         del_21765_6 == "del" &
         sample_date >= "2020-11-24" &
         lineage == "B.1.1.7"
        )
consortium_v
```
```{r}
mutations_s_uk %>% 
  summarise(n_distinct(variant))
```
```{r}
mutations_s_uk %>% 
  count(variant) %>% 
  filter(n == 1) 
```
```{r}
mutations_s_uk %>% 
  count(variant) %>% 
  filter(n >= 100) 
```

```{r}
mutations_s_uk %>% 
  count(variant) %>% 
  filter(n >= 5) %>% 
  arrange(desc(n))

```

### Table 1
```{r table_1}
database %>% 
  slice_max(`numSeqs UK`, n = 15) %>% 
  select(mutation, `numSeqs UK`, `numSeqs UK 28 days`, `numSeqs Eng 28 days`, `numSeqs Scotland 28 days`, `numSeqs Wales 28 days`, `numSeqs NI 28 days`)
```
```{r appendix}
database %>% 
  filter(`numSeqs UK` >= 5) %>% 
  arrange(desc(`numSeqs UK`)) %>% 
  select(mutation, `numSeqs UK`, `numSeqs UK 28 days`, `numSeqs Eng 28 days`, `numSeqs Scotland 28 days`, `numSeqs Wales 28 days`, `numSeqs NI 28 days`) %T>% write_csv("appendix.csv")
```


```{r}
setdiff(
consortium_uk %>% filter(d614g == "G") %>% select(sequence_name),
mutations_s_uk %>% filter(variant == "D614G") %>% select(sequence_name)
)
```
### Lineages
```{r lineage_counts}
n_uk_lineages <- 
  consortium_uk %>% 
  # filter(sample_date >= sample_date_28) %>% 
  group_by(lineage) %>%
  summarise(sequences = n(), 
            D614G = sum(d614g == "G"), 
            A222V = sum(a222v == "V"), 
            N439K = sum(n439k == "K"), 
            N501Y = sum(n501y == "Y"), 
            Y453F = sum(y453f == "F"),
            DEL_69_70 = sum(del_21765_6 == "del"), 
            N439K_DEL_69_70 = sum(n439k == "K" & del_21765_6 == "del"), 
            N501Y_DEL_69_70 = sum(n501y == "Y" & del_21765_6 == "del"),
            Y453F_DEL_69_70 = sum(y453f == "F" & del_21765_6 == "del")
  )

n_uk_lineages
```
#### B.1
```{r}
n_uk_lineages %>% 
  filter(lineage == "B.1" |str_detect(lineage, "^B\\.1\\.")) %>% 
  select(-lineage) %>% 
  summarise_all(funs(sum))
```

```{r}
n_uk_lineages %>% 
  filter(lineage == "B.1" |str_detect(lineage, "^B\\.1\\.")) %>% 
  mutate(lineage_top = "B.1") 
  
  
  
```
#### B.1.177
```{r}
n_uk_lineages %>% 
  filter(lineage == "B.1.177" | str_detect(lineage, "^B\\.1\\.177\\.")) %>% 
  select(-lineage) %>% 
  summarise_all(funs(sum))
```
#### B.1.141
```{r}
n_uk_lineages %>% filter(lineage == "B.1.141" | str_detect(lineage, "^B\\.1\\.141\\.")) %>% select(-lineage) %>% summarise_all(funs(sum))

```
#### B.1.258

```{r}
n_uk_lineages %>% filter(lineage == "B.1.258" | str_detect(lineage, "^B\\.1\\.258\\.")) %>% select(-lineage) %>% summarise_all(funs(sum))

```
#### B.1.1
```{r}
n_uk_lineages %>% filter(lineage == "B.1.1" | str_detect(lineage, "B\\.1\\.1\\.")) %>% select(-lineage) %>% summarise_all(funs(sum))
```

#### B.1.1.7
```{r}
n_uk_lineages %>% filter(lineage == "B.1.1.7" | str_detect(lineage, "B\\.1\\.1\\.7\\.")) %>% select(-lineage) %>% summarise_all(funs(sum))
```


#### B.1.1.70
```{r}
n_uk_lineages %>% filter(lineage == "B.1.1.70" | str_detect(lineage, "B\\.1\\.1\\.70\\.")) %>% select(-lineage) %>% summarise_all(funs(sum))

```

#### B.1.351
```{r}
n_uk_lineages %>% filter(lineage == "B.1.351" | str_detect(lineage, "B\\.1\\.351\\.")) %>% select(-lineage) %>% summarise_all(funs(sum))
```

#### B.1.1.298
```{r}
n_uk_lineages %>% filter(lineage == "B.1.1.298" | str_detect(lineage, "B\\.1\\.1\\.298\\.")) %>% select(-lineage) %>% summarise_all(funs(sum))
```
#### E484K
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  count(lineage)
```
#### E484K 28 Days
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(sample_date >= sample_date_28) %>% 
  count(lineage)
```
#### E484K / B.1.351 Summary
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(lineage == "B.1.351" | str_detect(lineage, "B\\.1\\.1\\.351\\.")) %>% 
  summarise("n_E484K_B.1.351" = n_distinct(sequence_name))
```
#### E484K / B.1.351 Summary 28 Days
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(lineage == "B.1.351" | str_detect(lineage, "B\\.1\\.1\\.351\\.")) %>% 
  filter(sample_date >= sample_date_28) %>% 
  summarise("n_28_E484K_B.1.351 " = n_distinct(sequence_name))
```
#### E484K / B.1.1.248
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(lineage == "B.1.1.248" | str_detect(lineage, "B\\.1\\.1\\.248\\.")) %>% summarise("n_E484K_B.1.1.248" = n_distinct(sequence_name))

```
#### E484K / B.1.1.248 28 Days
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(lineage == "B.1.1.248" | str_detect(lineage, "B\\.1\\.1\\.248\\.")) %>% 
  filter(sample_date >= sample_date_28) %>% 
  summarise("n_E484K_B.1.1.248" = n_distinct(sequence_name))
```

#### E484K / B.1.1.7
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(lineage == "B.1.1.7" | str_detect(lineage, "B\\.1\\.1\\.7\\.")) %>% summarise("n_E484K_B.1.1.7" = n_distinct(sequence_name))
```
#### E484K / B.1.1.7 28 Days
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(lineage == "B.1.1.7" | str_detect(lineage, "B\\.1\\.1\\.7\\.")) %>% 
  filter(sample_date >= sample_date_28) %>% 
  summarise("n_E484K_B.1.1.7" = n_distinct(sequence_name))
```

#### E484K / B.1.1
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(lineage == "B.1.1" | str_detect(lineage, "B\\.1\\.1\\.")) %>% summarise("n_E484K_B.1.1" = n_distinct(sequence_name))
```
#### E484K / B.1.1 28 Days
```{r}
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(sample_date >= sample_date_28) %>% 
  filter(lineage == "B.1.1" | str_detect(lineage, "B\\.1\\.1\\.")) %>% summarise("n_E484K__28_B.1.1" = n_distinct(sequence_name))
```
#### N501Y + E484K / B.1.351 Summary
```{r}
intersect(
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(lineage == "B.1.351" | str_detect(lineage, "B\\.1\\.1\\.351\\.")) %>% 
  distinct(sequence_name),
  
consortium_uk %>% 
  filter(n501y == "Y") %>% 
  filter(lineage == "B.1.351" | str_detect(lineage, "B\\.1\\.1\\.351\\.")) %>% 
  distinct(sequence_name)
)
```

#### E484K + N501Y / B.1.351
```{r}
intersect(
mutations_s_uk %>% 
  filter(variant == "E484K") %>% 
  filter(lineage == "B.1.351" | str_detect(lineage, "B\\.1\\.1\\.351\\.")) %>% 
  filter(sample_date >= sample_date_28) %>% 
  distinct(sequence_name),
  
consortium_uk %>% 
  filter(n501y == "Y") %>% 
  filter(lineage == "B.1.351" | str_detect(lineage, "B\\.1\\.1\\.351\\.")) %>% 
  filter(sample_date >= sample_date_28) %>% 
  distinct(sequence_name)
)
```
### Escape Mutations
```{r escape}
escape <- 
  read_csv("list_escape_2.csv", col_names = "mutation")

database %>% 
  right_join(escape) %>% 
  select(mutation, `numSeqs UK`:`numSeqs Wales 28 days`) %T>% 
  write_csv("escape_counts.csv")
```

### Escape mutations sample names
```{r escape_ids}
escape_t4 <- 
c(
"G446V",
"L452R",
"E484Q",
"K444R",
"E484K",
"Y508H",
"N440K",
"L455F",
"A831V",
"A475V",
"F490S",
"V483A",
"R346K",
"K378N",
"K444N",
'G446S',
"N450D",
"K150R",
"R346S",
"K150T",
"V445A",
"G446A",
"Y449H",
"E484A",
"Y453F", "N439K"
)  

lapply(escape_t4,  function(x){
    mutations_s_uk %>% 
        filter(variant == x) %>% 
        distinct(sequence_name) %T>% {{.} %$% 
        write_lines(sequence_name, paste0(x, ".txt"))}
})
```
## nsp12
```{r nsp12}
mutations_uk %>% 
  mutate(across(position, as.character)) %>% 
  mutate(across(position, as.numeric)) %>% 
  filter(gene == "orf1ab" & between(position, 4393, 5324)) %>% 
  mutate(nsp12_variant = str_c(str_sub(variant, 1, 1), position - 4392, str_sub(variant, -1, -1) ) ) %>% 
  group_by(gene, variant, nsp12_variant) %>% 
  summarise(sequences = n_distinct(sequence_name)) %>% 
  ungroup %>% 
  slice_max(sequences, n = 15) %T>% write_csv("nsp12_top15.csv")
```

