library(shiny)
library(DT)
library(shinyWidgets)
library(shinydashboard)
library(shinydashboardPlus)
library(shinyjs)
library(formattable)
library(plotly)

dashboardPage(
    title = "COG-UK Mutation Explorer",
    skin = "blue",
    
    header = dashboardHeader(title = tags$a(href='http://cogconsortium.uk', target = "_blank",
                                   tags$img(src='image2.png', height = "50px"))),
    
    sidebar = dashboardSidebar(
        sidebarMenu(id="sidebar_menu",
            menuItem("Mutation Explorer", tabName = "report", icon = icon("table")), 
            menuItem("Visualiser", tabName = "dashboard", icon = icon("virus")),
            menuItem("Antigenic Information", tabName = "immunology", icon = icon("shield-virus")),
            menuItem("About", tabName = "about", icon = icon("info-circle"))
        ),
        
        conditionalPanel(condition =  "input.sidebar_menu == 'dashboard'",
                         selectInput("gene", "Gene:", mutations_uk %>% distinct(gene) %>% arrange(gene), 
                                     selected = "S"),
                         selectInput("position", "Position:", 
                                     mutations_uk %>% distinct(position) %>% arrange(position), 
                                     selected = "614", selectize = TRUE),
                         prettySwitch("percentage", "Percentage", FALSE, status = "info"),
                         prettySwitch("ref", "Wild type / other", TRUE,  status = "info"),
                         
                         prettyRadioButtons(
                             inputId = "nation",
                             label = "UK nation:", 
                             choices = c("UK", "England", "Northern Ireland" = "Northern_Ireland", "Scotland", "Wales"),
                             inline = FALSE, 
                             status = "info",
                             fill = TRUE,
                             selected = "UK"
                         ),
                         
                         chooseSliderSkin("Modern"),
                         setSliderColor("#5bc0de", 1), # Bootstrap info colour
                         sliderTextInput(
                             inputId = "epi_week",
                             label = "Epidemic week range:", 
                             choices = consortium_uk$epi_week %>% levels,
                             selected = c(consortium_uk$epi_week %>% levels %>% first, 
                                          consortium_uk$epi_week %>% levels %>% last),
                             animate = TRUE
                         )
        )
    ),
    
    body = dashboardBody(
        useShinyjs(), # set up the dashboard to use shinyjs  
        tabItems(
            tabItem(tabName = "about",
                    fluidRow(
                        h1("COG-UK / Mutation Explorer"),
                        p(str_c("The COG-UK / Mutation Explorer (COG-UK/ME) provides information and structural context on mutations and associated variants in the genes encoding SARS-COV-2 proteins that have been identified from sequence data generated by the COVID-19 Genomics (COG-UK) Consortium by ", dataset_date %>% format("%A %d %B %Y.")),"We focus on SARS-CoV-2 spike gene mutations of potential or known importance based on epidemiological, clinical and/or experimental observations."),
                        p("The Mutation Explorer comprises of:"),
                        tags$ol(
                            tags$li("high frequency individual amino acid replacements (Table 1), a subset of which may be important"),
                            tags$li("mutations of potential or known clinical and public health importance based on current evidence (Table 2)"),
                            tags$li("the designated global variants of concern and their structural context (Table 3)"),
                            tags$li("frequency plots for mutations at specific residue in SARS-CoV-2 ORFs (Visualiser)"),
                            tags$li("mutations of potential antigenic significance as indicated by experimental studies: shown to lead to weaker neutralisation of the virus by convalescent plasma from people who have been infected with SARS-CoV-2 and/or demonstrated escape from some monoclonal antibodies (mAbs) that may be given to patients with COVID-19 (Antigenic Information).")
                        ),
                        p(em("Notes"), " provides an explanation of scientific terms and a few limitations."),
                        
                        h3("Credits"),
                        p("COG-UK/ME is developed within and funded by the COVID-19 Genomics UK Consortium by Derek W. Wright, Joseph Hughes, William Harvey, Ben Jackson, Andrew Rambaut, David L. Roberson, Alessandro M. Carabelli. COG-UK/ME is based on the CLIMB framework, and maintained by the MRC-University of Glasgow Centre for Virus Research. Please see the 'How to cite' page if you intend to use data from this web site. CLIMB data provided via this web application is subject to CLIMB Terms and Conditions. Contact", a(href = "mailto:Derek.Wright@glasgow.ac.uk", "Derek.Wright@glasgow.ac.uk"), "with questions or feedback. Follow", a(href ="https://twitter.com/CovidGenomicsUK", target = "_blank", "COG-UK"), "to be notified of updates.")
                    )),
            tabItem(tabName = "report",
                fluidRow(
                    h1("COG-UK / Mutation Explorer"),
                    h3("Latest UK sequence:", max(consortium_uk$sample_date) %>% format("%A %d %B %Y")),
                    
                    tabBox(
                        title = "Analysis", width = 12,
                        id = "tabs_report",
                        tabPanel("Table 1", 
                                 h3("1. Spike gene mutations"),
                                 p("Individual amino acid replacements detected in UK
                                 genomes are shown in Table 1. Amino acid
                                 replacements detected in 5 or more complete SARS-CoV-2 genomes in the COG-UK dataset are included, ranked by the frequency of the replacement. Neither insertions nor deletions are included."),
                                 
                                 p(em("NB Number of genomes is not equal to number of COVID-19 cases as data have not been deduplicated.")),
                                 dataTableOutput("table_1"), 
                                 
                                 h4("Download data"),
                                 p("Download a CSV file, for each amino acid replacement, comprising COG-UK sequence name, sample date, epidemic week, global lineage, UK lineage and phylotype. UK sequences are filtered by a 28 day period up to and including the most recent UK sequence date."), 
                                 selectInput("dataset", "Choose amino acid replacement:",
                                             choices = 
                                                 database %>% 
                                                 filter(`numSeqs UK` >= 5 & `numSeqs UK 28 days` > 0) %>% 
                                                 arrange(desc(`numSeqs UK`)) %$% 
                                                 mutation, 
                                             selectize = FALSE),
                                 downloadButton("downloadData", "Download", class = "btn-info")
                        ),
                        tabPanel("Table 2", 
                                 h3("2. Spike gene mutations of potential importance"),
                                 p("Single spike gene mutations of potential or clinical and public health importance based on
                                 current evidence are listed in Table 2."),
                                 
                                 h5("CAVEATS:"),
                                 tags$ul(
                                     tags$li("The table aims to provide information on individual mutations, but this is rapidly
                                     becoming an over-simplification because mutations are increasingly arising in a
                                     range of combinations."),
                                     
                                     tags$li("The term ‘escape’ is used in the table as shorthand to mean weaker neutralisation of
                                     the virus by convalescent plasma from people who have been infected with SARS-
                                         COV-2, and/or some monoclonal antibodies (mAbs) that may be given to patients
                                     with COVID-19.")
                                 ),
                                 
                                 h4("Table 2. Spike S gene mutations, lineage associations and reason for interest, UK
                                 lineages"),
                                 p(em("NB Numbers are lower than in Table 1 because Table 2 only considers specific lineages.")),
                                 tableOutput("table_2")
                        ), 
                        tabPanel("Table 3",
                                 h3("3. Global variants of concern being monitored by UK PHAs"),
                                 tableOutput("table_3"),
                                 
                                 h4("Download data"),
                                 p("Download a CSV file containing COG-UK sequence name, sample date, epidemic week, global lineage, UK lineage and phylotype. Cumulative UK sequences are filtered by the selected lineage of concern."), 
                                 selectInput("concern", "Choose lineage:",
                                             choices = c(lineages_t3$lineage, "B.1.1.7 + E484K") %>% sort, # TODO exclude zero count lineages
                                             selectize = FALSE),
                                 downloadButton("downloadConcern", "Download", class = "btn-info")
                        ), 
                        tabPanel("Notes",                         
                                 h2("Data source and processing"),
                                 p(str_c("The analysis described in this report is based on ", n_distinct(consortium_uk$sequence_name) %>% comma(format = "d")
                                   ," UK-derived genomes sequenced by COG-UK: complete data in the MRC-CLIMB database to ", dataset_date %>% format("%d/%m/%Y"), ", with the latest sequence from ", max(consortium_uk$sample_date) %>% format("%d/%m/%Y"),  ".")),
                                 p("A report of the geographic distribution and prevalence of SARS-CoV-2 lineages in general, and global variants of interest, can be found ", a(href = "https://cov-lineages.org/global_report.html", target = "_blank", "here", .noWS = "outside"), ". Amino acid replacement, insertion and deletion counts for all SARS-CoV-2 genes in the global GISAID database can be found ", a(href = "http://cov-glue.cvr.gla.ac.uk/", target = "_blank", "here", .noWS = "outside"), ".", .noWS = c("after-begin", "before-end")),
                                 h2("Limitations"),
                                 tags$ol(
                                     tags$li("This report is for information only. The clinical and public health importance of any single mutation, or combination of mutations cannot be determined from sequence data alone."),
                                     tags$li("Putative evidence for the importance of any single mutation, or combination of mutations can be derived from computational biology and further evaluated by laboratory experiments. Genomic and laboratory evidence then need to be combined with clinical datasets that are designed to allow detection of increased transmissibility, change in disease severity, drug resistance or altered vaccine efficacy. For this reason, surveillance and risk assessment of mutations and variants is a multi-agency process involving UK Public Health Agencies who have access to detailed information on patients and populations, and other groups including NERVTAG (New and Emerging Respiratory Virus Threats Advisory Group)."),
                                     tags$li("COG-UK generates around 10,000 genomes a week, which will rise to 20,000 per week by March 2021. When COVID-19 infection rates are high, not all viruses from infected people will be sequenced and some mutations at low frequency will not be detected, but COG-UK aims to take representative samples from across the UK.")
                                 ),
                                 
                                 h2("Appendix 1"),
                                 h3("Background"),
                                 p("Mutations arise naturally in the SARS-CoV-2 genome as the virus replicates and circulates in the
                    human population. As a result of this on-going process, many thousands of mutations have already
                    arisen in the SARS-CoV-2 genome since the virus emerged in late 2019. As mutations continue to
                    arise, novel combinations of mutations are increasingly observed. The vast majority of mutations have
                    no apparent effect on the virus. Only a very small minority are likely to be important and change the
                    virus in any appreciable way. This could include a change in the ability to infect/transmit between
                    people; a change in disease severity; or a change in the way the virus interacts with the immune
                    system (including the response generated by a vaccine). We pay most attention to mutations in the
                    gene that encodes the Spike protein, which is associated with viral entry into cells and it is relevant in
                    the context of immunity and vaccine efficacy."),
                                 h3("Definitions"),
                                 tags$ul(
                                     tags$li(em("Mutation"), " is used to describe a change of a nucleotide in the virus RNA genome, a subset of which
                        results in a change in amino acid (sometimes referred to as a substitution or replacement), or a
                        mutation can refer to a deletion or insertion event in the virus genome. By convention an amino acid
                        change is written N501Y to denote the wildtype (N, asparagine) and replacement amino acid (Y, 
                        tyrosine) at site 501 in the amino acid sequence."),
                                     
                                     tags$li(em("Viral variant"), " refers to a genetically distinct virus with different mutations to other viruses. Variant can
                        also refer to the founding virus of a cluster/lineage and used to refer collectively to the resulting
                        variants that form the lineage."),
                                     
                                     tags$li(em("Lineages"), " are assigned combining genetic and, in the case of SARS-CoV-2 due to weak phylogenetic
                        signals, also with epidemiological data. COG-UK uses the nomenclature system introduced by
                        Rambaut et al. (2020), see https://cov-lineages.org."),
                                     
                                     tags$li(em("VUI"), " is used by Public Health England to indicate Variant Under Investigation."),
                                     
                                     tags$li(em("VOC"), " is used by Public Health England to indicate Variant of Concern. Lineage B.1.1.7 was initially
                        named by Public Health England as VUI 202012/01 (Variant Under Investigation, year 2020, month
                        12, variant 01) and subsequently redesignated as VOC-202012/01 (Variant of Concern 202012/01).")
                                 )
                        )
                        
                    ) # end tabBox
                ) # end fluidRow
            ), # end tabItem
            
            tabItem(tabName = "dashboard",
                    fluidRow(
                        box(plotlyOutput("mutation_time", height = 748), width = 12)
                    )
            ),
            
            tabItem(tabName = "immunology",
                    h2("Antigenic Information"),
                    h3("Spike protein gene mutations of potential immunogenic significance significance detected in the UK"),
                    p('The table lists those mutations in the spike gene identified in the UK dataset that have been
                                 associated with weaker neutralisation of the virus by convalescent plasma from people who
                                 have been infected with SARS-CoV-2, and/or some mAbs that may be given to patients with
                                 COVID-19 (referred to below as "escape").'),
                    p(strong("There is no evidence at the time of writing for this impacting on the efficacy of current
                                 vaccines or the immune response to natural SARS-CoV-2 infection.")),
                    
                    h4('Reported "escape" mutations in the spike gene detected in the UK'),
                    DTOutput("table_4"),
                    
                    h4("Download data"),
                    p("Download a CSV file containing COG-UK sequence name, sample date, epidemic week, global lineage, UK lineage and phylotype. Cumulative UK sequences are filtered by the selected amino acid replacement."), 
                    selectInput("escape", "Choose amino acid replacement:",
                                choices = database %>% 
                                    filter(!is.na(escape)) %>% 
                                    filter(`numSeqs UK` > 0) %>% 
                                    arrange(desc(`numSeqs UK`), mutation) %$% 
                                    mutation, 
                                selectize = FALSE),
                    downloadButton("downloadEscape", "Download", class = "btn-info")
            )
        ) # end tabItems
    ), # end dashboardBody
    
    footer = dashboardFooter(
        left = fluidRow(
            column(3, tags$a(img(src = "CVR.png", width = "205px", height = "71px", class = "img-responsive center-block"), href="https://www.gla.ac.uk/researchinstitutes/iii/cvr/", target = "_blank")),
            column(3, tags$a(img(src = "UoG_colour.png", width = "229px", height = "71px", class = "img-responsive center-block"), href="https://www.gla.ac.uk/", target = "_blank")),
            column(3, tags$a(img(src = "CLIMB.png", width = "234px", height = "71px", class = "img-responsive center-block"), href="https://www.climb.ac.uk/", target = "_blank")),
        column(3, tags$a(img(src = "UOE.png", width = "294px", height = "71px", class = "img-responsive center-block"), href="https://www.ed.ac.uk", target = "_blank")))
     )
)